### PR TITLE
Rework map.jinja according to the template-formula and add support for Debian 10

### LINKED ISSUE #108: Debian 10 support with tomcat9
I'm using this formula for Debian 9 (tomcat8) and wanted to use it for Debian 10 as well (tomcat9),  which isn't supported by the formula yet. At first that seemed as easy as applying the correct information to the codenamemap. 

However, the map.jinja seems to merge the osmap and oscodenamemap in the reverse order (at least compared to the implementation in most formulas). 
In most formulas the map.jinja works with mapping yml files, where the more generic mappings are overruled by the more specific mappings. E.g.:
  defaults.yml -> osfamilymap.yml -> osmap.yml -> oscodenamemap.yml

Is there a good reason why this is implemented differently in this fomula's map.jinja? As it doesn't seem to work as expected.

Are there objections against changing that in a more common used?

I can create a pull request if that works.

### PR DESCRIPTION
**Pull request by danny-smit**
This change implements issue #108.

Changes:
- Add support for Debian 10  (tomcat 9)
- Reworked map.jinja according to the [template-formula](https://github.com/saltstack-formulas/template-formula) to solve an ordering issue of the os specifics.

I tried to run the current tests in order to extend them, but they seem to be broken already (or I'm doing it wrong).

So a thorough review might be used.

### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [ ] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Related issues and/or pull requests

Close #108.

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

File: tomcat/map.jinja [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
+{%- set config = salt['grains.filter_by'](
+      {'defaults': defaults},
+      default='defaults',
+      merge=_config
+    )
+%}
 
-{# merge the actual tomcat pillar into the above combined dict #}
-{% set tomcat = salt['pillar.get']('tomcat', default=lookup, merge=True) %}
+{#- Change **TEMPLATE** to match with your formula's name and then remove this line #}
```
Comment (noelmcloughlin) [inline tomcat/map.jinja:39]: This line is not needed.

File: tomcat/osfamilymap.yaml [hunk 2]
```diff
@@ -14,14 +14,15 @@ Debian:
   user: tomcat8
   group: tomcat8
   java_home: /usr/lib/jvm/default-java
+  # Use OS defaults:
+  catalina_base:
```
Comment (noelmcloughlin) [inline tomcat/osfamilymap.yaml:18]: Use empty string for yaml . `catalina_base: ''`

**Comment by myii** [comment]
@danny-smit I've implemented `semantic-release` in #110, since there were a number of issues I could see with this formula.  By adding testing to this formula, that also has made it easier to evaluate this PR.  To save some time around the changes I've got for this PR, are you OK if I rebase your PR, add the minor changes and push it here?

**Comment by danny-smit** [comment]
@myii That would be great, rebasing is fine with me.

File: tomcat/map.jinja (hunk 1 above)
Comment (danny-smit) [reply tomcat/map.jinja]: Thanks, I removed the line.

File: tomcat/osfamilymap.yaml (hunk 2 above)
Comment (danny-smit) [reply tomcat/osfamilymap.yaml]: Thanks, I pushed new changes according to your comment.

**Comment by noelmcloughlin** [comment]
> LGTM. thank you for the PR.

There are conflicts to be fixed.

**Comment by danny-smit** [comment]
I rebased my changes on top of the master to resolve the conflicts.

**Comment by danny-smit** [comment]
The tests are failing on the catalina_* variables. Is there a reason why they must be hard-coded in the defaults.yaml? Relying on the defaults of the OS seemed to work fine for several OS versions.

The downside of that is that you are obligated to include them for every osfamily and os version that deviates from the default value.

**Comment by noelmcloughlin** [comment]
> Relying on the defaults of the OS seemed to work fine for several OS versions.

Yes, Software defaults should have precedence - formula should not try to be smarter.  This repo has no maintainer so feel free to impose your own interpretation of a good solution as long as it's reasonable.

**Comment by myii** [comment]
I didn't get around to merging this after pushing the changes.  I added the rest of the standard structure from the `template-formula` as well, for easier long-term updates, now that this is a `semantic-release` formula.

@danny-smit @noelmcloughlin I removed the change to those `catalina` variables for _this PR_, so that we keep this refactor as pure as possible.  Let's tackle that in a separate PR where required.  A few considerations for this formula:

* As with all coding situations, changes made should be consistent with the rest of the code around it; it's inconsistent to have the `catalina` settings provided for some/most of the platforms and not for the rest.  The settings that were already present were being applied across _all_ Debian platforms, so we need to consider the impact.  If this situation only applies to `debian-10`, it can be modified in `osfingermap.yaml` instead.
* Don't overlook the issues this raises when writing tests.  If software is being configured _implicitly_, it becomes increasingly difficult to write tests _explictly_ -- it makes it really difficult to figure out which paths to check.  If formulas clearly define where installations/configurations are going to take place, the tests can be written duplicating the same values.
* While this formula has no _explicit_ maintainer, it's still being maintained by the core team in general -- so things must be kept as clear as possible, to avoid making the maintenance burden even more difficult.  But if someone is willing to put their name down for a formula (i.e. using the `CODEOWNERS` solution), the situation could be reconsidered.

Thanks to you both for the contribution and the reviews!

**Comment by danny-smit** [comment]
@myii Thanks, well done getting test tests in order! I can live with your choice if that is the preferred choice, as longs as it works. 

However, the current changes in master do not work out-of-the-box on my machines in a real life situation. On Debian 9 the tomcat8 service fails to start with an obscure message saying:
```
tomcat8[6190]: tomcat8 is not installed ... failed!
```

On Debian 10 all states finish successfully, but the tomcat9 service ends up in a failed state with the error:
```
tomcat9[3808]: /usr/libexec/tomcat9/tomcat-start.sh: 25: cd: can't cd to /usr/share/tomcat
```

If I manually remove the CATALINA_* variables from the 'main_config' file, the services are starting normally again. So the current values do not seem to work for Debian hosts, possible also not for Ubuntu. Debian typically appends the tomcat (major) version number to the 'tomcat' directories in those paths.

This doesn't seem to be included in the tests, because the init.sls currently only installs and starts tomcat. It isn't handling the configuration.

**Comment by myii** [comment]
> This doesn't seem to be included in the tests, because the init.sls currently only installs and starts tomcat. It isn't handling the configuration.

@danny-smit Thanks for continuing the discussion.  This formula was a mess before I implemented `semantic-release`, I'm surprised it worked on anything:

* https://travis-ci.org/github/myii/tomcat-formula/builds/676728028

So this is a process of improvement, starting with an initial structure that verifiably works.  Now we can refine that to improve the other bits that don't work.  It's just a matter of time investment.

Since you've got a clearly defined problem there, let's attempt to tackle that using the framework we've got in place now.  I should have some feedback fairly soon, I just need a little while.

**Comment by myii** [comment]
@danny-smit For example, just tested this using Kitchen locally for `debian-10` and is building fine:

```yaml
Debian-10:
  ...
  catalina_base: /usr/share/tomcat9
  catalina_home: /usr/share/tomcat9
  catalina_pid: /var/run/tomcat9.pid
  catalina_tmpdir: /var/cache/tomcat9/temp
```

Would you be able to confirm that adding those four lines to `osfingermap.yaml` work at your end?

Ultimately, users can use empty values for these values in pillar/config to prevent them being set.

**Comment by danny-smit** [comment]
These are the values that work for Debian-10 (catalina_base is different):
```
Debian-10:
  ...
  catalina_base: /var/lib/tomcat9           
  catalina_home: /usr/share/tomcat9         
  catalina_pid: /var/run/tomcat9.pid        
  catalina_tmpdir: /var/cache/tomcat9/temp  
```
I tried the same for Debain 9 with tomcat8, but the values didn't work yet. I'll come back to that.

It would be nice if we can add working values, otherwise new users will run into the same issues if they are not aware if this.

**Comment by danny-smit** [comment]
For Debian-9 I used the following:
```yaml
Debian-10:
  ...
  catalina_base: /var/lib/tomcat8           
  catalina_home: /usr/share/tomcat8         
  catalina_pid: /var/run/tomcat8.pid        
  catalina_tmpdir: /var/cache/tomcat8/temp  
```
However the tomcat8 service doesn't like the `catalina_tmpdir` because it doesn't exist by default (even though permission are correct for the service to create it by itself).
The following possible solutions seem to work:
- ensuring the directory exists with the proper permissions
- using a different path, by default the jvm_tmp path is selected on Debian-9:
  - /tmp/tomcat8-tomcat8-tmp   # default of the OS
  - /tmp/tomcat   # default specified by jvm_tmp in the defaults.yaml

**Comment by myii** [comment]
@danny-smit Thanks for your responses.

> These are the values that work for Debian-10 (catalina_base is different):
> 
> ```yaml
> Debian-10:
>   ...
>   catalina_base: /var/lib/tomcat9           
>   ...
> ```

Interesting, where did you derive that `catalina_base` from?  Myself, I took the sample values that were already present in `/etc/default/tomcat9` and adjusted them where necessary.  But I've just tested with your value and it also works fine at my end for `debian-10`.

> I tried the same for Debain 9 with tomcat8, but the values didn't work yet. I'll come back to that.
> 
> It would be nice if we can add working values, otherwise new users will run into the same issues if they are not aware if this.

Sure, we can resolve those as we go along.  We can at least get `debian-10` up and running.

> For Debian-9 I used the following:
> 
> ...
> 
> However the tomcat8 service doesn't like the `catalina_tmpdir` because it doesn't exist by default (even though permission are correct for the service to create it by itself).
> The following possible solutions seem to work:
> 
> * ensuring the directory exists with the proper permissions
> * using a different path, by default the jvm_tmp path is selected on Debian-9:
>       
>       * /tmp/tomcat8-tomcat8-tmp   # default of the OS
>       * /tmp/tomcat   # default specified by jvm_tmp in the defaults.yaml

So we have options here.  We could use `jvm_tmp path` if `catalina_tmpdir` isn't defined (or has been cleared via. the pillar/config).  Either way, we can add a [`file.directory`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.directory) state to ensure everything is in place with the right permissions.

**Comment by danny-smit** [comment]
> > These are the values that work for Debian-10 (catalina_base is different):
> > ```yaml
> > Debian-10:
> >   ...
> >   catalina_base: /var/lib/tomcat9           
> >   ...
> > ```
> 
> Interesting, where did you derive that `catalina_base` from? Myself, I took the sample values that were already present in `/etc/default/tomcat9` and adjusted them where necessary. But I've just tested with your value and it also works fine at my end for `debian-10`.

I performed a clean installation of tomcat9. After that there were no `CATALINA_*` variables defined in my `/etc/default/tomcat9`.

Therefore I looked further to find the default values and found this in the service file (/lib/systemd/system/tomcat9.service):
```sh
Environment="CATALINA_HOME=/usr/share/tomcat9"
Environment="CATALINA_BASE=/var/lib/tomcat9"
Environment="CATALINA_TMPDIR=/tmp"
```

For Debian-9 I found it in /etc/init.d/tomcat8 (snippet with the relevant variables):
```sh
NAME=tomcat8
JVM_TMP=/tmp/tomcat8-$NAME-tmp
CATALINA_HOME=/usr/share/$NAME
CATALINA_BASE=/var/lib/$NAME
CATALINA_PID="/var/run/$NAME.pid"

if [ -z "$CATALINA_TMPDIR" ]; then
        CATALINA_TMPDIR="$JVM_TMP"
fi
```

> So we have options here. We could use `jvm_tmp path` if `catalina_tmpdir` isn't defined (or has been cleared via. the pillar/config). Either way, we can add a [`file.directory`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.directory) state to ensure everything is in place with the right permissions.

Yes, adding a `file.directory` state seems like the most robust solution, which should work even in case a user defines a different path using the pillar/config.

Do you want me to create a new pull request? Or are you following up on this yourself?

**Comment by myii** [comment]
@danny-smit Please, feel free to add a PR.  @noelmcloughlin has already swamped me with about 15 massive PRs to work through!

**Comment by myii** [comment]
@danny-smit Nice explanation, by the way.  For the `file.directory`, somewhere before or after this block seems appropriate:

https://github.com/saltstack-formulas/tomcat-formula/blob/71616297f4c4301ac7b901f36df1528c49552890/tomcat/config.sls#L8-L29

**Pull request merged**